src/error.rs

```diff
@@ -6,37 +6,43 @@
 // KIND, either express or implied. Please review the Licences for the specific language governing
 // permissions and limitations relating to use of the SAFE Network Software.
 
+use crate::{messages, section};
 use thiserror::Error;
 
 /// The type returned by the sn_routing message handling methods.
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
-/// Internal error.
+/// Routing error.
+// TODO: consider removing those variants that cannot occur when calling the public API.
 #[derive(Debug, Error)]
 #[allow(missing_docs)]
 pub enum Error {
-    #[error("Failed signature check.")]
-    FailedSignature,
-    #[error("Cannot route.")]
-    CannotRoute,
-    #[error("Network layer error: {}", .0)]
+    #[error("network layer error: {}", .0)]
     Network(#[from] qp2p::Error),
-    #[error("The node is not in a state to handle the action.")]
-    InvalidState,
-    #[error("Bincode error: {}", .0)]
-    Bincode(#[from] bincode::Error),
-    #[error("Invalid source location.")]
+    #[error("node bootstrap was interrupted")]
+    BootstrapInterrupted,
+    #[error("failed to create genesis section: {}", .0)]
+    CreateGenesisSection(#[source] section::CreateError),
+    #[error("failed to merge sections: {}", .0)]
+    MergeSection(#[from] section::MergeError),
+    #[error("failed to create message: {}", .0)]
+    CreateMessage(#[from] messages::CreateError),
+    #[error("received invalid message: {}", .0)]
+    InvalidMessage(#[from] messages::IntegrityError),
+    #[error("failed to send a message")]
+    SendMessage,
+    #[error("failed to create vote proof share: {}", .0)]
+    CreateVoteProofShare(#[source] bincode::Error),
+    #[error("received invalid vote proof share")]
+    InvalidVoteProofShare,
+    #[error("failed to create resource challenge: {}", .0)]
+    CreateResourceChallenge(#[source] bincode::Error),
```

Comment:
Maybe for `CreateResourceChallenge` and `SendMessage`, it shall includ a detailed reason of detailed error? like what `CreateMessage` does ?

Reply:
Agree about `SendMessage`, but the resource challenge creation can fail only due to serialization.

Reply:
Ok.
Then just change for SendMessage